File: java/DynamicMP.java
import java.lang.Math;
import java.util.LinkedList;
import java.util.TreeMap;

/*

implementation of dynamic data structure for maintaining a coreset

*/

// our dynamic coreset data structure
public class DynamicMP {

  // the nested collections of points for each layer
  private LinkedList<TreeMap<Integer, float[]>> U;

  // the sampled medians for each layer
  private LinkedList<TreeMap<Integer, Integer>> S;

  // the clusters for each layer
  private LinkedList<TreeMap<Integer, Integer>[]> C;

  // parameters for layer construction
  private int k;

  private float alpha;
  private float beta;

  private int sampleSize;

  public DynamicMP(int k, float alpha, float beta) {

    // parameters
    this.k = k;
    this.alpha = alpha;
    this.beta = beta;

    // sampling parameter
    this.sampleSize = (int)Math.floor(alpha*k);
  }

  // insert a point into the data structure
  public void insert(int key, float[] data) {

  }

  // delete a point from the data structure
  public void delete(int key) {

  }

}

File: python/dynamicMP.py
import math
import numpy as np

# implementation of dynamic version of Mettu-Plaxton's static k-medians algorithm

# our dynamic coreset algoithm, based on a dynamization of MP
class DynamicMP:

    # initialize the data structure
    def __init__(self, d, k, alpha, beta):

        # the metric
        self.d = d

        # parameters
        self.k = k
        self.alpha = alpha
        self.beta = beta

        # sampling paramter
        self.sampleSize = floor(self.alpha * self.k)

    # class to define a layer of the dynamic MP algorithm
    class Layer:

        # construct layer
        def __init__(self, U):

            # the metric space to be clustered
            self.U = U

            # the size of the metrix space
            self.n = len(self.U)

            # sample floor(\alpha * k) elements from U u.a.r with replacement
            samples = np.random.randint(self.sampleSize, size=self.n)

        # insert a point into the layer
        def insert(x):

            return None

        # delete a point in the layer
        def delete(x):

            return None

    # insert a point into the data structure
    def insert(x):

        return None

    # delete a point from the data structure
    def delete(x):

        return None

print(115)
